### PR TITLE
Improve container security

### PR DESCRIPTION
**Pull request by sashayakovtseva**
This PR adds two major changes: 
1) user to run container process is validated and set correctly
2) overlayfs is prepared in a slightly different way to enable suid-flow inside a container. To achieve this upper and work directories are created in common directory that is remounted without nosuid flag later.

### CODE REVIEW

File: pkg/kube/container_files.go [hunk 1]
```diff
@@ -91,20 +94,22 @@ func (c *Container) addOCIBundle(image *image.Info) error {
...
+	return nil
 }
 
 func (c *Container) prepareOverlay(imagePath string) error {
 	var (
+		suidDir   = filepath.Join(c.bundlePath(), "suid")
 		lowerPath = filepath.Join(c.bundlePath(), "lower")
-		upperPath = filepath.Join(c.bundlePath(), "upper")
-		workPath  = filepath.Join(c.bundlePath(), "work")
+		upperPath = filepath.Join(suidDir, "upper")
```
Comment (cclerget) [inline pkg/kube/container_files.go]: I find `suidDir` a bit confusing, `overlayDir` seems more appropriate

File: pkg/kube/container_files.go [hunk 2]
```diff
@@ -91,20 +94,22 @@ func (c *Container) addOCIBundle(image *image.Info) error {
...
 	)
 
-	log.Printf("creating %s", lowerPath)
-	err := os.Mkdir(lowerPath, 0755)
+	// prepare upper and work directories
+	log.Printf("creating %s", suidDir)
+	err := os.Mkdir(suidDir, 0755)
 	if err != nil {
-		return fmt.Errorf("could not create lower directory for overlay: %v", err)
+		return fmt.Errorf("could not create suid directory for overlay: %v", err)
```
Comment (cclerget) [inline pkg/kube/container_files.go]: `could not create overlay parent directory`

File: pkg/kube/container_files.go [hunk 3]
```diff
@@ -116,21 +121,35 @@ func (c *Container) prepareOverlay(imagePath string) error {
 	if err != nil {
 		return fmt.Errorf("could not create working directory for overlay: %v", err)
 	}
-
-	log.Printf("creating %s", c.rootfsPath())
-	err = os.Mkdir(c.rootfsPath(), 0755)
+	err = syscall.Mount(suidDir, suidDir, "", syscall.MS_BIND, "")
+	if err != nil {
+		return fmt.Errorf("could not bind mount suid directory: %v", err)
```
Comment (cclerget) [inline pkg/kube/container_files.go]: `could not bind mount overlay parent directory`

File: pkg/kube/container_files.go [hunk 4]
```diff
@@ -116,21 +121,35 @@ func (c *Container) prepareOverlay(imagePath string) error {
...
+	}
+	err = syscall.Mount(suidDir, suidDir, "", syscall.MS_REMOUNT|syscall.MS_BIND, "")
 	if err != nil {
-		return fmt.Errorf("could not create root directory for overlay: %v", err)
+		return fmt.Errorf("could not remount suid directory: %v", err)
```
Comment (cclerget) [inline pkg/kube/container_files.go]: `could not remount overlay parent directory`

File: pkg/kube/container_files.go [hunk 5]
```diff
@@ -91,20 +94,22 @@ func (c *Container) addOCIBundle(image *image.Info) error {
 	if err != nil {
 		return fmt.Errorf("could not encode OCI config into json: %v", err)
 	}
-	return c.prepareOverlay(image.Path())
+	return nil
 }
 
 func (c *Container) prepareOverlay(imagePath string) error {
 	var (
+		suidDir   = filepath.Join(c.bundlePath(), "suid")
```
Comment (cclerget) [inline pkg/kube/container_files.go]: `overlay` ?

File: pkg/kube/container_oci.go [hunk 6]
```diff
@@ -322,3 +321,50 @@ func (t *containerTranslator) configureAnnotations() {
...
+	if err != nil {
+		return err
+	}
+
+	t.g.SetProcessUID(uint32(containerUser.Uid))
+	t.g.SetProcessGID(uint32(containerUser.Gid))
+	for _, gid := range containerUser.Sgids {
+		t.g.AddProcessAdditionalGid(uint32(gid))
+	}
+	for _, gid := range security.GetSupplementalGroups() {
```
Comment (cclerget) [inline pkg/kube/container_oci.go]: Additional groups are not checked ?

File: pkg/kube/container_oci.go [hunk 7]
```diff
@@ -322,3 +321,50 @@ func (t *containerTranslator) configureAnnotations() {
...
+
+	userSpec := strings.Join(userParts, ":")
+	containerUser, err := getContainerUser(t.cont.rootfsPath(), userSpec)
+	if err != nil {
+		return err
+	}
+
+	t.g.SetProcessUID(uint32(containerUser.Uid))
+	t.g.SetProcessGID(uint32(containerUser.Gid))
+	for _, gid := range containerUser.Sgids {
```
Comment (cclerget) [inline pkg/kube/container_oci.go]: Does it take groups associated with user from container `/etc/group` ?

File: pkg/kube/container_oci.go (hunk 6 above)
Comment (sashayakovtseva) [reply pkg/kube/container_oci.go]: No, they are simply added. Only primary group is validated (if present)

File: pkg/kube/container_oci.go (hunk 7 above)
Comment (sashayakovtseva) [reply pkg/kube/container_oci.go]: Yes. `containerUser` is already a valid user. Upon validation container's `/etc/passwd` and `/etc/groups` are read to make sure user & group exist.

**Pull request merged**